### PR TITLE
Swc 2298

### PR DESCRIPTION
**Pull request by Njanderson**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/view/ProfileViewImpl.java [hunk 1]
```diff
@@ -589,11 +589,12 @@ public void setIsMoreProjectsVisible(boolean isVisible) {
...
 	
-	private void addProjectBadges(List<ProjectHeader> projectHeaders, FlowPanel targetPanel) {
+	private void addProjectBadges(List<ProjectHeader> projectHeaders, List<UserProfile> lastModifiedBy, FlowPanel targetPanel) {
 		//uses ProjectBadge to show more information (additional info available from ProjectHeader)
-		for (ProjectHeader projectHeader : projectHeaders) {
+		GWT.debugger();
+		for (int i = 0; i < projectHeaders.size(); i++) {
 			ProjectBadge badge = ginInjector.getProjectBadgeWidget();
-			badge.configure(projectHeader);
+			badge.configure(projectHeaders.get(i), lastModifiedBy == null ? null :lastModifiedBy.get(i));
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/client/view/ProfileViewImpl.java]: Move logic to presenter. Make an "addProjectBadge"

File: src/main/java/org/sagebionetworks/web/client/view/ProfileViewImpl.java [hunk 2]
```diff
@@ -563,8 +563,8 @@ public void setTeamsError(String error) {
 	}
 	
 	@Override
-	public void addProjects(List<ProjectHeader> projectHeaders) {
-		addProjectBadges(projectHeaders, projectsTabContent);
+	public void addProjects(List<ProjectHeader> projectHeaders, List<UserProfile> modifiedBy) {
+		addProjectBadges(projectHeaders, modifiedBy, projectsTabContent);
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/client/view/ProfileViewImpl.java]: singular -> plural: modifiedByList

File: src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadge.java [hunk 3]
```diff
@@ -69,40 +71,49 @@ public Widget asWidget() {
...
 		order.add("ID");
 		map.put("ID", header.getId());
 
-		order.add("Modified By");
-		map.put("Modified By", modifiedBy);
-		
-		order.add("Modified On");
+		if (modifiedBy != null) {
+			order.add("Modified By");
+			map.put("Modified By", modifiedBy.getFirstName() + " " + modifiedBy.getLastName()
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadge.java]: DisplayUtils getDisplayName instead

File: src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadge.java [hunk 4]
```diff
@@ -69,40 +71,49 @@ public Widget asWidget() {
...
+//		}
+//		UserBadge.getUserProfile("" + header.getModifiedBy(), adapterFactory, synapseClient, clientCache, new AsyncCallback<UserProfile>() {
+//			@Override
+//			public void onSuccess(UserProfile profile) {
+//				if (view.isAttached())
+//					callback.onSuccess(
+//							
+//							profileToKeyValueDisplay(profile, DisplayUtils.getDisplayName(profile)));		
+//			}
+//
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadge.java]: Delete comment

File: src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadgeView.java [hunk 5]
```diff
@@ -26,6 +27,8 @@
 	 * Presenter interface
 	 */
 	public interface Presenter {
-		void getInfo(final AsyncCallback<KeyValueDisplay<String>> callback);
+//		void getInfo(final AsyncCallback<KeyValueDisplay<String>> callback);
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadgeView.java]: Delete comment

File: src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadge.java [hunk 6]
```diff
@@ -69,40 +71,49 @@ public Widget asWidget() {
 		return view.asWidget();
 	}	
 	
-	public KeyValueDisplay<String> profileToKeyValueDisplay(UserProfile profile, String modifiedBy) {
+	@Override
+	public KeyValueDisplay<String> profileToKeyValueDisplay() {
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadge.java]: Pass tooltip formatted text onto the view

File: src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadgeViewImpl.java [hunk 7]
```diff
@@ -86,27 +86,38 @@ public void setLastActivityText(String text) {
...
-						if (isPopover)
-							tooltip.show();
-					}
-				}
-			});
+			KeyValueDisplay<String> result = presenter.profileToKeyValueDisplay();
+			String content = ProvViewUtil.createEntityPopoverHtml(result).asString();
+			isPopoverInitialized = true;
+			if (widget.isAttached()) {
+				tooltip.setTitle(content);
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadgeViewImpl.java]: In view configure, configure with tooltip text

File: src/test/java/org/sagebionetworks/web/unitserver/SynapseClientImplTest.java [hunk 8]
```diff
@@ -1740,6 +1741,7 @@ public void testGetMyProjects() throws Exception {
 		verify(mockSynapse).getMyProjects(eq(ProjectListType.MY_PROJECTS),
 				eq(ProjectListSortColumn.LAST_ACTIVITY),
 				eq(SortDirection.DESC), eq(limit), eq(offset));
+		verify(mockSynapse).listUserProfiles(anyList());
 	}
```
Comment (Njanderson) [inline src/test/java/org/sagebionetworks/web/unitserver/SynapseClientImplTest.java:1747]: Add test for key-value string map creation

File: src/main/java/org/sagebionetworks/web/shared/ProjectPagedResults.java [hunk 9]
```diff
@@ -37,43 +54,5 @@ public void setTotalNumberOfResults(int totalNumberOfResults) {
 	public void setResults(List<ProjectHeader> results) {
 		this.results = results;
 	}
-
-	
-
-
-	@Override
-	public int hashCode() {
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/shared/ProjectPagedResults.java:45]: Forgot/not sure what to do for this, equals, and toString.

File: src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadge.java [hunk 10]
```diff
@@ -69,39 +72,25 @@ public Widget asWidget() {
 		return view.asWidget();
 	}	
 	
-	public KeyValueDisplay<String> profileToKeyValueDisplay(UserProfile profile, String modifiedBy) {
+	@Override
+	public String getProjectTooltip() {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/ProjectBadge.java:81]: test

File: src/main/java/org/sagebionetworks/web/client/presenter/ProfilePresenter.java [hunk 11]
```diff
@@ -466,14 +475,30 @@ public void onFailure(Throwable caught) {
 		});
 	}
 	
-	public void addProjectResults(List<ProjectHeader> headers) {
+	public void addProjectResults(List<ProjectHeader> projectHeaders, List<UserProfile> lastModifiedByList) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/ProfilePresenter.java:477]: test

File: src/main/java/org/sagebionetworks/web/client/presenter/ProfilePresenter.java [hunk 12]
```diff
@@ -466,14 +475,30 @@ public void onFailure(Throwable caught) {
...
+			widget.addStyleName("margin-bottom-10 col-xs-12");
+			view.addProjectWidget(widget);
+		}
+		if (projectHeaders.isEmpty())
+			view.addProjectWidget(new HTML(SafeHtmlUtils.fromSafeConstant("<div class=\"smallGreyText padding-15\">" + EntityTreeBrowserViewImpl.EMPTY_DISPLAY + "</div>").asString()));
 	}
 	
 	public void addChallengeResults(List<ChallengeBundle> challenges) {
 		view.showChallengesLoading(false);
-		view.addChallenges(challenges);
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/ProfilePresenter.java:474]: test

File: src/main/java/org/sagebionetworks/web/shared/ProjectPagedResults.java [hunk 13]
```diff
@@ -37,43 +54,5 @@ public void setTotalNumberOfResults(int totalNumberOfResults) {
...
-		final int prime = 31;
-		int result = 1;
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/shared/ProjectPagedResults.java:47]: regenerate

**Pull request merged**